### ISSUE TITLE
Not working: SSL/TLS without WebSockets

### ISSUE THREAD

**Issue by johnjore** [eclipse/mosquitto#194]
Unable to connect with certificates if WebSockets is not enabled.

Used M2Mqtt C# library and the MQTT.fx client.

Certs are signed using internal CA. Certs work when WebSockets is enabled. Tested with OwnTracks app.

As M2Mqtt does not support WebSockets, having mosquitto working with certs, but without WebSockets, would be very usefull...

JJ

**Comment by ralight** [comment]
Could you post a config file that shows this behaviour please? And to make my life easier, could you also confirm whether using mosquitto_sub as the client does the same thing?

**Comment by johnjore** [comment]
Hi! I've replaced the host/domain names, IP addresses and the filenames, the rest is kept as is.

conf.conf file from conf.d folder:
> listener 8881 hostname.domain.xx
> cafile   /etc/mosquitto/certs/root.cer
> certfile /etc/mosquitto/certs/host.cer
> keyfile  /etc/mosquitto/certs/nopwd.key
> require_certificate true
> use_identity_as_username false
> allow_anonymous false
> 
> listener 8883 1.1.1.1
> cafile   /etc/mosquitto/certs/root.cer
> certfile /etc/mosquitto/certs/host.cer
> keyfile  /etc/mosquitto/certs/nopwd.key
> protocol websockets
> require_certificate true
> use_identity_as_username true
> allow_anonymous false
> 
> listener 8882 1.1.1.1
> require_certificate false
> use_identity_as_username false
> allow_anonymous false

mosquiotto.conf (No changes made)
> # Place your local configuration in /etc/mosquitto/conf.d/
> # A full description of the configuration file is at
> # /usr/share/doc/mosquitto/examples/mosquitto.conf.example
> log_type error
> log_type warning
> log_type notice
> log_type information
> log_type debug
> log_type subscribe
> log_type unsubscribe
> connection_messages true
> log_timestamp true
> log_dest syslog
> log_dest stdout
> log_dest topic
> log_dest file /var/log/mosquitto/mosquitto.log
> 
> autosave_interval 1800
> 
> persistence true
> persistence_location /var/lib/mosquitto/
> persistence_file mosquitto.db
> 
> pid_file /var/run/mosquitto.pid
> include_dir /etc/mosquitto/conf.d
> 

When I comment out the "protocol websocket" line, the Android app OwnTracks shows this:
<img width="270" alt="screenshot_2016-06-27-18-21-39" src="https://cloud.githubusercontent.com/assets/12951842/16389085/36faf7f0-3c94-11e6-9c01-f7d2617c6023.png">
When WebSocket is re-enabled in the config file, it connect without issues. No changes to the certs or key files. I'll see if I can figure out the syntax for mosquitto_sub.

JJ

**Comment by ralight** [comment]
I can't figure out what could be happening.

I've tried to reproduce this with the certificates in the test directory but it's all ok. There is an asciicast of me doing it here:

**Comment by ralight** [comment]
I'm closing this for now, if you get more information please reopen it.

**Comment by johnjore** [comment]
Hi! Got sidetracked, but I've tried to reproduce by doing a git clone to get the same files:

Config file using the same certs:
> listener 18881
> cafile   /src/mosquitto/test/ssl/all-ca.crt
> certfile /src/mosquitto/test/ssl/server.crt
> keyfile  /src/mosquitto/test/ssl/server.key
> require_certificate true
> use_identity_as_username false
> allow_anonymous false
> 
> listener 18883
> cafile   /src/mosquitto/test/ssl/all-ca.crt
> certfile /src/mosquitto/test/ssl/server.crt
> keyfile  /src/mosquitto/test/ssl/server.key
> protocol websockets
> require_certificate true
> use_identity_as_username false
> allow_anonymous false

The asciicast cuts the right hand side, but by trying all the filenames:

> ls /src/mosquitto/test/ssl/test-*.crt -laFh
> -rw-r--r-- 1 root root 3.0K Jul 15 12:56 /src/mosquitto/test/ssl/test-alt-ca.crt
> -rw-r--r-- 1 root root 1005 Jul 15 12:56 /src/mosquitto/test/ssl/test-bad-root-ca.crt
> -rw-r--r-- 1 root root  997 Jul 15 12:56 /src/mosquitto/test/ssl/test-fake-root-ca.crt
> -rw-r--r-- 1 root root  997 Jul 15 12:56 /src/mosquitto/test/ssl/test-root-ca.crt
> -rw-r--r-- 1 root root 3.0K Jul 15 12:56 /src/mosquitto/test/ssl/test-signing-ca.crt
> :/etc/mosquitto/test# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/test-alt-ca.crt --key /src/mosquitto/test/ssl/client.key
> Unable to connect (A TLS error occurred.).
> :/etc/mosquitto/test# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/test-bad-root-ca.crt --key /src/mosquitto/test/ssl/client.key
> Unable to connect (A TLS error occurred.).
> :/etc/mosquitto/test# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/test-fake-root-ca.crt --key /src/mosquitto/test/ssl/client.key
> Unable to connect (A TLS error occurred.).
> :/etc/mosquitto/test# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/test-root-ca.crt --key /src/mosquitto/test/ssl/client.key
> Unable to connect (A TLS error occurred.).
> :/etc/mosquitto/test# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/test-signing-ca.crt --key /src/mosquitto/test/ssl/client.key
> Unable to connect (A TLS error occurred.).

> OpenSSL> version
> OpenSSL 1.0.1t  3 May 2016

> 1468607936: mosquitto version 1.4.9 (build date Fri, 03 Jun 2016 09:02:12 +0100) starting

So... Why do my tests fail with the TLS error, but you have success?


JJ

**Comment by ralight** [comment]
The asciicast uses tmux to be able to show two things at once - so that you can see the broker messages at the same time as connecting a client.

Looks like you managed to try every combination apart from the one that should work :)

Try:

    mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/client.crt --key /src/mosquitto/test/ssl/client.key

**Comment by johnjore** [comment]
:) and :(

> :/src/mosquitto# mosquitto_sub -p 18883 --cafile /src/mosquitto/test/ssl/test-root-ca.crt -t asdf --cert /src/mosquitto/test/ssl/client.crt --key /src/mosquitto/test/ssl/client.key
> Error: A TLS error occurred.

?

**Comment by johnjore** [comment]
Resolved. 

So, long story short, nothing wrong with my root cert or the cert request, issue is related to how the default Windows CA "Web server" profile signes the cert.

I happened to have created a slightly different signing profile for VMware, using this profile to sign the request has finally resolved the issue.  I can't remember exactly what was changed for the VMWare certs, but there are a few recepies out there.

Still, no idea why the original certs work in WebSocket mode, but not without WebSocket.

JJ
